File: Python/dynamic_programming/coin_change.py
def get_count(coins, amt):
    def _get_count(coins, n, amt):
        dp = [[None] * (n + 1) for _ in range(amt + 1)]

        for i in range(n + 1):
            dp[0][i] = 1
        for i in range(amt + 1):
            dp[i][0] = 0
        for i in range(1, amt + 1):
            for j in range(1, n + 1):
                dp[i][j] = dp[i][j - 1]
                if coins[j - 1] <= i:
                    dp[i][j] += dp[i - coins[j - 1]][j]
        return dp[amt][n]

    return _get_count(coins, len(coins), amt)


if __name__ == "__main__":
    print(get_count([2, 3, 5, 6], 10))

File: Java/dynamicprogramming/CoinChange.java
package dynamicprogramming;

public class CoinChange {

  /**
   * Time Complexity => O(amount * no. of coins)
   * <p>
   * Space Complexity => O(amount * no. of coins)
   * 
   * @param coins  array of coins
   * @param amount targeted amount
   * @return no. of ways to get the targeted amount using the given coins
   */
  public int getCount(int[] coins, int amount) {
    return coinsCount(coins, coins.length, amount);
  }

  private int coinsCount(int[] coins, int n, int amount) {
    int[][] dp = new int[amount + 1][n + 1];

    for (int i = 0; i <= n; i++)
      dp[0][i] = 1;
    for (int i = 0; i <= amount; i++)
      dp[i][0] = 0;

    for (int i = 1; i <= amount; i++)
      for (int j = 1; j <= n; j++) {
        dp[i][j] = dp[i][j - 1];
        if (coins[j - 1] <= i)
          dp[i][j] += dp[i - coins[j - 1]][j];
      }
    return dp[amount][n];
  }

  public static void main(String[] args) {
    CoinChange ob = new CoinChange();
    System.out.println(ob.getCount(new int[] { 2, 3, 5, 6 }, 10));
  }
}
